bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractBoschSHCHandlerTest.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.boschshc.internal.devices;
+
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.openhab.binding.boschshc.internal.devices.bridge.BridgeHandler;
+import org.openhab.binding.boschshc.internal.exceptions.BoschSHCException;
+import org.openhab.core.config.core.Configuration;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingStatusInfo;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandlerCallback;
+
+/**
+ * Abstract unit test implementation for all types of handlers.
+ *
+ * @author David Pace - Initial contribution
+ *
+ * @param <T> type of the handler to be tested
+ */
+@ExtendWith(MockitoExtension.class)
+public abstract class AbstractBoschSHCHandlerTest<T extends BoschSHCHandler> {
+
+    private T fixture;
```

Comment:
When I annotate this class with `@NonNullByDefault` I have the following problem: some members (like in this case `fixture`)  are considered as not always non-null by the compiler. In all other cases i can fix this by annotating the member with `@NonNullByDefault({})`.

The problem is that this only seems to work if the field has a concrete type. With generics (like `T`) in this case this does not work. Do you know a solution for that?

Reply:
I can have a look probably tomorrow, I seem to remember having experienced something similar, but don't remember the details.

Reply:
You could try to keep `fixture` non-null and add a constructor:
```java
    public AbstractBoschSHCHandlerTest() {
        fixture = createFixture();
    }
```

Reply:
That works, thanks 😃

I'm not 100% convinced by this solution because now the fixures are created twice (once in the constructor and once in `beforeEach()`), and the ones created in the constructor are in invalid states because the mocks are not initialized yet. But since the fixtures that are created in `beforeEach()` are valid and it works, I would be fine with this solution.

Reply:
Yeah, I'm also not fully satisfied with that hack for the same reason. At least this is test code, so performance should not be much of a concern in this specific case. The same trick is sometimes being used in handlers also because `initialize()` is too late to satisfy the static code analysis.